staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -459,31 +459,9 @@ func (cnc *CloudNodeController) syncNode(ctx context.Context, nodeName string) e
 	})
 
 	err = clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
-		var curNode *v1.Node
-		if cnc.cloud.ProviderName() == "gce" {
-			// TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
-			// Since there are node taints, do we still need this?
-			// This condition marks the node as unusable until routes are initialized in the cloud provider
-			if err := nodeutil.SetNodeCondition(cnc.kubeClient, types.NodeName(nodeName), v1.NodeCondition{
-				Type:               v1.NodeNetworkUnavailable,
-				Status:             v1.ConditionTrue,
```

Comment:
Not sure if this is still safe to remove, I think this was added due to a similar check we have in kubelet to add this condition before initialization. It's possible this check is no longer needed though but I'm misisng context on it. @jprzychodzen might know.

Reply:
at this point the node is tainted, see line 427

```
	cloudTaint := getCloudTaint(curNode.Spec.Taints)
	if cloudTaint == nil {
		// Node object received from event had the cloud taint but was outdated,
		// the node has actually already been initialized, so this sync event can be ignored.
		return nil
	}
```


and the route controller will set the corresponding condition if necessary

https://github.com/kubernetes/kubernetes/blob/52457842d155743f0e3fc57ade87251cca37d375/staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go#L336-L371

Reply:
> and the route controller will set the corresponding condition if necessary 

Not sure I follow, but the condition is here _because_ node controller is about to remove the taint, allowing workloads to schedule on this node. The condition here ensures pods won't schedule without route controller doing a second pass on the node.

Even with route controller applying the condition, there is still a window of time between when nodes are schedulable and the condition is applied. Note that route controller does not watch Nodes and instead does a full sync of all nodes every 5 minutes by default. So in the worse case scenario you'll have a 5 minute window of pods being scheduled without networking. This is also why kubelet applied the condition as well, to ensure pods are not scheduable until Node is synced by route controller at least once.

Reply:
but this is already added in the kubelet, no?

https://github.com/kubernetes/kubernetes/blob/52457842d155743f0e3fc57ade87251cca37d375/pkg/kubelet/kubelet_node_status.go#L336-L345

Reply:
I think the https://github.com/kubernetes/kubernetes/blob/52457842d155743f0e3fc57ade87251cca37d375/pkg/kubelet/kubelet_node_status.go#L337 may cause some issues after all.

Based on https://github.com/kubernetes/kubernetes/blob/52457842d155743f0e3fc57ade87251cca37d375/pkg/kubelet/kubelet_network.go#L30-L40

The `kubelet.cloud` should be nil - using the `external` cloudprovider which should be the generic case now - and therefor the condition is not set. However @andrewsykim I must ask and please correct me, but wouldn't such a scenario that you describe be applicable to any cloudprovider/setup that rely on the route controller and is not named `gce` ? And if so, shouldn't we strive for a common solution instead of relying on special/legacy "gceism"s as the PR title eloquently put ? 

From the previous point, maybe kubelet should always set the condition then. For overlay networks Calico/Cilium are already managing the condition and for other cases that include the route controller, we know that it is already updating the condition. I am not familiar enough to comment about the practices of other overlay-supporting CNIs though. 

Maybe a kubelet configuration flag as the comments should be the way forward such that it controls whether the kubelet is setting the condition on node creation is more flexible.

Reply:
> may cause some issues after all.

ohh, you are right, I missed that, sorry 

> And if so, shouldn't we strive for a common solution instead of relying on special/legacy "gceism"s as the PR title eloquently put ?

heh, don't take the title too serious :) , unfortunately there are some things in the code like that we can not break, I will try to come with a solution that does not break people and that solve problems like yours, this one does not seem to be good to go

/hold

Reply:
@aojea It was not my intention to sound sarcastic or any of such sort! Rather I honestly think that @andrewsykim 's scenario is valid and we could make a generic improvement. 

Regarding the problem at hand, I believe that allowing control over the initial `NetworkUnavailable` condition on the kubelet is rather beneficial. With cloudprovider `external`, the kubelet would rather have a hard time with heuristics to determine what should be the initial condition settings, so delegate that to the cluster provider/owner. With something like that, your PR makes more sense (but would require some adaptations from `gce` users to their kubelet config).

Having said that, I understand that this is complex to navigate and not break things along. Thank you for looking into it.

Reply:
to close on this, Dan perfectly summarizes it here https://github.com/kubernetes/kubernetes/issues/120486#issuecomment-1710410069 , the network plugin on the nodes, if it depends on the PodCIDR fields to be populated, should not mark the node as ready.
In GCE this is already done by netd https://github.com/GoogleCloudPlatform/netd/blob/df4eab3811e20430b21ec8ec9113b407c93a6cb7/scripts/install-cni.sh#L131 , so this taint is redundant, as the node will not be ready until the cni config file is present